manager/network/cni/cni.go

```diff
@@ -0,0 +1,84 @@
+package cni
+
+import (
+	"github.com/containernetworking/cni/libcni"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/allocator/cniallocator"
+	"github.com/docker/swarmkit/manager/network"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+)
+
+type cni struct{}
+
+// New produces a fresh CNI Network Model
+func New() network.Model {
+	return &cni{}
+}
+
+func (nm *cni) NewAllocator() (network.Allocator, error) {
+	return cniallocator.New()
+}
+
+func (nm *cni) SupportsIngressNetwork() bool {
+	return false
+}
+
+func parseCNIspec(spec *api.NetworkSpec) (*libcni.NetworkConfig, error) {
+	// This is rather similar to cniConfig in the containerd executor...
+	cniConfig, ok := spec.DriverConfig.Options["config"]
+	if !ok {
+		return nil, grpc.Errorf(codes.InvalidArgument, "CNI network has no config")
+	}
+
+	cni, err := libcni.ConfFromBytes([]byte(cniConfig))
+	if err != nil {
+		return nil, grpc.Errorf(codes.InvalidArgument, "Failed to parse CNI config: %s", err)
+	}
+	return cni, nil
+}
+
+func (nm *cni) SetDefaults(spec *api.NetworkSpec) error {
+	if spec.DriverConfig == nil || spec.DriverConfig.Name != "cni" {
+		return nil
+	}
+
+	cni, err := parseCNIspec(spec)
+	if err != nil {
+		return err
+	}
+
+	if spec.Annotations.Name == "" {
+		spec.Annotations.Name = cni.Network.Name
```

Comment:
This could have far reaching implications, for example places like `validateAnnotations` would need a special case which would depend on the type of spec the annotation was embedded in and we would need to arrange that uses of the network spec always got their name via a wrapper function instead of going to it directly.

There's also the overhead of having to parse the JSON each time (although I have an open plan to try caching that).

I was hoping to avoid requiring the user to specify both `spec.Annotations.Name` and the cni name identically, but perhaps I should flip this around and instead allow the cni name to be empty and fill it in from the (required) annotation name. That would "just" require me to figure out how to round trip an arbitrary JSON object through a parser to add a field without knowing in advance all of the fields which are present. I'm not actually sure if the go json library can do that, but I'll investigate.

Or another alternative would be to add a `Name` field to the `api.Network` object which contains the canonical name as determined by looking at the spec and the cni. AIUI that object is, unlike the spec, internal and fair game for mutating.

WDYT?